# 目录

   * [RNA-seq分析](#rna-seq分析)
      * [0. 介绍](#0-介绍)
      * [1. 前期准备](#1-前期准备)
      * [2. 工具下载](#2-工具下载)
         * [2.0 生信管理工具](#20-生信管理工具)
         * [2.1 sratoolkit](#21-sratoolkit)
         * [2.2 fastqc](#22-fastqc)
         * [2.3 multiqc](#23-multiqc)
         * [2.4 cutadapt](#24-cutadapt)
         * [2.5 trimmomatic](#25-trimmomatic)
         * [2.6 hisat2](#26-hisat2)
         * [sortmerna](#sortmerna)
         * [2.7 samtools](#27-samtools)
         * [2.8 HTseq](#28-htseq)
         * [2.9 R](#29-r)
         * [2.10 Rstudio](#210-rstudio)
         * [2.11 parallel](#211-parallel)
         * [StringTie[可选]](#stringtie可选)
         * [Ballgown[可选]](#ballgown可选)
      * [3. 数据下载](#3-数据下载)
         * [3.1 参考数据](#31-参考数据)
            * [Ensembl](#ensembl)
         * [3.2 测试数据（实验数据）](#32-测试数据实验数据)
      * [4. 质量控制](#4-质量控制)
         * [4.1 质量评估](#41-质量评估)
         * [4.2 剔除接头以及测序质量差的碱基](#42-剔除接头以及测序质量差的碱基)
         * [4.4 再次去除低质量区域](#44-再次去除低质量区域)
         * [4.3 再次查看质量情况](#43-再次查看质量情况)
      * [5. 去除rRNA序列](#5-去除rrna序列)
      * [6. 序列比对](#6-序列比对)
         * [6.1 建立索引](#61-建立索引)
         * [6.2  开始比对](#62--开始比对)
      * [7. 表达量统计](#7-表达量统计)
      * [8. 合并表达矩阵与标准化](#8-合并表达矩阵与标准化)
         * [8.1 合并](#81-合并)
         * [8.2 数据标准化](#82-数据标准化)
            * [8.2.1 简介](#821-简介)
         * [8.2.2 cufflinks](#822-cufflinks)
         * [8.2.3 手动计算](#823-手动计算)
      * [9. 差异表达分析](#9-差异表达分析)
         * [9.1 数据前处理](#91-数据前处理)
         * [9.2 差异分析](#92-差异分析)
            * [9.2.1 安装与加载包](#921-安装与加载包)
            * [9.2.2 构建对象](#922-构建对象)
            * [9.2.3 样本相关性](#923-样本相关性)
            * [9.2.4 差异基因](#924-差异基因)
      * [10. 提取差异表达基因与注释](#10-提取差异表达基因与注释)
         * [10.1 名词解释](#101-名词解释)
         * [10.2 使用ClusterProfiler对基因的ID进行转化](#102-使用clusterprofiler对基因的id进行转化)
         * [10.3 使用biomaRt进行注释](#103-使用biomart进行注释)
      * [11. 可视化](#11-可视化)
      * [12. 富集分析](#12-富集分析)
         * [12.1 Gene Ontology (GO)分析](#121-gene-ontology-go分析)
         * [12.2 KEGG分析](#122-kegg分析)
         * [12.3 GSEA分析](#123-gsea分析)
         * [12.4 DO（Disease Ontology）分析](#124-dodisease-ontology分析)
         * [12.4 另外可以使用几个在线网站](#124-另外可以使用几个在线网站)
      * [========================================](#-1)
      * [5. 表达量分析](#5-表达量分析)
      * [6. 表达量分析](#6-表达量分析)
      * [7. 差异表达分析](#7-差异表达分析)
      * [作者](#作者)
      * [参考](#参考)
         * [流程](#流程)
         * [结果解读](#结果解读)
         * [原理](#原理)
         * [程序下载安装](#程序下载安装)
         * [问题](#问题)

# RNA-seq分析

## 0. 介绍
```
转录组（Transcriptome）被定义为一个细胞中存在的所有转录本读取的集合。
RNA-seq数据用于研究和/或量化一个生物体的转录组，它可用于以下类型的实验：

差异基因表达（Differential Gene Expression）: 转录本水平的定量评估和比较
转录组组装（Transcriptome assembly）: 构建基因组转录区域的轮廓，是一种定性评估
能被用于帮助建立更好的基因模型，并使用组装来验证
元转录组学或群体转录组学分析
RNA-seq 除了得到差异表达的基因之外，还可以对SNP、新颖的转录本、可变剪接、RNA编辑等进行分析，但是最为常见的是得到差异表达的基因。
```

各种软件功能简介：
```
     database                   Workflow                        tools
======================================================================================
  
+=================+     +-------------------------+                               
|     database    |     |      Quality Analysis   |---------------> fastqc 
+=================+     +-------------------------+                                
|+------+         |                 v                                      
|| rRNA |---------|--+  +-------------------------+
|+------+         |  |  | Base Quality Filtering  |------------> TrimGalore
|  +------+       |  |  +-------------------------+
|  |genome|-------|-+|              v
|  +------+       | ||  +-------------------------+
|     +----------+| |+->| rRNA Sequence Filtering |------------> SortMeRNA
|     |  Genome  || |   +-------------------------+
|     |Annotation|| |               v
|     +----------+| |   +-------------------------+
|          |      | +-->|   Genome Alignment      |------------> hisat2
+----------|------+     +-------------------------+
           |                        v
           |            +-------------------------+
           +----------->|  Count Mapped Reads     |------------> HTseq
                        +-------------------------+
                                    v
                        +-------------------------+
                        | Differential Expression |------------> DESeq2
                        +-------------------------+
                                    v
                        +-------------------------+
                        |     Pathway analysis    |------------> ClusterProfiler
                        +-------------------------+
```
各类数据分析步骤及需要用到的软件：  
* RNA-seq  
① MultiQC 整理数据，general statistics  
② featureCounts： Subread featureCounts是一个高效的通用阅读摘要程序，用于计算基因组特征（如基因，外显子，启动子，基因体，基因组箱和染色体位置）的映射读取。    
③ cutadapt： Cutadapt是一种工具，用于从高通量测序读取中查找和删除adapter序列，引物，poly-A尾部和其他类型的不需要的序列。   
④ fastQC: FastQC是高通量序列数据的质量控制工具。  
* whole-genome seq :   
① MultiQC 整理数据，general statistics  
② QualiMap： 是一个独立于平台的应用程序，用于促进比对序列数据及其衍生物（如特征计数）的质量控制。        
③ SnpEff： SnpEff是一个遗传变异注释和效果预测工具箱。它注释和预测变异对基因的影响（如氨基酸变化）。     
④ GATK: GATK是一个工具包，提供各种工具，主要关注变异发现和基因分型。  
⑤ Picard： Picard 是一组用于操作高通量测序数据的 Java 命令行工具。  
⑥ FastQ Screen ：FastQ Screen允许您根据一组序列数据库筛选FastQ格式的序列库，以便您可以查看库的组成是否与您期望的匹配。  
⑦ fastQC: FastQC是高通量序列数据的质量控制工具。   
* Bisulfite seq   
① MultiQC 整理数据，general statistics  
② Bismark ： Bismark是一种绘制亚硫酸氢盐转化序列读数和确定胞嘧啶甲基化状态的工具。  
③ cutadapt： Cutadapt是一种工具，用于从高通量测序读取中查找和删除adapter序列，引物，poly-A尾部和其他类型的不需要的序列。  
④ fastQC: FastQC是高通量序列数据的质量控制工具。  
* Hi-C  
① MultiQC 整理数据，general statistics  
② HICUP： HiCUP（Hi-C user pipeline）是一种用于对Hi-C数据进行映射和执行质量控制的工具。  
③ fastQC: FastQC是高通量序列数据的质量控制工具。   
* MultiQC  
① MultiQC 整理数据，general statistics  
② edgeR：样本相似性由通过 edgeR 归一化的基因计数生成。然后计算 log2 归一化 CPM 值之间的欧氏距离并进行聚类。  
③ MDS PLOT: 显示项目中样本之间的相关性。这些值是在 edgeR_heatmap_MDS.r 脚本中使用 edgeR 计算的。    
④ STAR: STAR是一种超快的通用RNA-seq aligner。  
⑤ cutadapt： Cutadapt是一种工具，用于从高通量测序读取中查找和删除adapter序列，引物，poly-A尾部和其他类型的不需要的序列。  
⑥ fastQC: FastQC是高通量序列数据的质量控制工具。    






## 1.前期准备
将目录建在d盘 cd /mnt/d  
建立目录    
`mkdir biosoft`   
`mkdir -p project/rat`    
`mkdir annotation genome sequence output script`  
 
结果：  
```
xuruizhi@DESKTOP-HI65AUV:/mnt/d/project/rat$ tree
.
├── annotation
├── genome
├── output
├── script
└── sequence

5 directories, 0 files
```

## 2.工具下载 
### 2.0 生信管理工具 

Linux brew  
来源[wang-q Ubuntu -](https://github.com/wang-q/ubuntu#install-linuxbrew)

### 2.1 sratoolkit   
* 使用brew安装  
```
xuruizhi@DESKTOP-HI65AUV:~$ brew install sratoolkit

HOMEBREW_BREW_GIT_REMOTE set: using https://mirrors.tuna.tsinghua.edu.cn/git/homebrew/brew.git for Homebrew/brew Git remote.
HOMEBREW_CORE_GIT_REMOTE set: using https://mirrors.tuna.tsinghua.edu.cn/git/homebrew/homebrew-core.git for Homebrew/core Git remote.
Running `brew update --auto-update`...
Warning: sratoolkit 3.0.0 is already installed and up-to-date.
To reinstall 3.0.0, run:
  brew reinstall sratoolkit
```
### 2.2 fastqc  
* 使用brew安装  
```
xuruizhi@DESKTOP-HI65AUV:~$ brew install fastqc
HOMEBREW_BREW_GIT_REMOTE set: using https://mirrors.tuna.tsinghua.edu.cn/git/homebrew/brew.git for Homebrew/brew Git remote.
HOMEBREW_CORE_GIT_REMOTE set: using https://mirrors.tuna.tsinghua.edu.cn/git/homebrew/homebrew-core.git for Homebrew/core Git remote.
Running `brew update --auto-update`...
Warning: fastqc 0.11.9_1 is already installed and up-to-date.
To reinstall 0.11.9_1, run:
  brew reinstall fastqc
```
### 2.3 multiqc 
``` 
# 使用python的安装器安装
pip install multiqc
```
```
xuruizhi@DESKTOP-HI65AUV:~$ pip install multiqc
DEPRECATION: Configuring installation scheme with distutils config files is deprecated and will no longer work in the near future. If you are using a Homebrew or Linuxbrew Python, please see discussion at https://github.com/Homebrew/homebrew-core/issues/76621
Requirement already satisfied: multiqc in /home/linuxbrew/.linuxbrew/lib/python3.9/site-packages (1.12)
Requirement already satisfied: jinja2>=2.9 in /home/linuxbrew/.linuxbrew/lib/python3.9/site-packages (from multiqc) (3.1.2)
Requirement already satisfied: click in /home/linuxbrew/.linuxbrew/lib/python3.9/site-packages (from multiqc) (8.1.3)
Requirement already satisfied: future>0.14.0 in /home/linuxbrew/.linuxbrew/lib/python3.9/site-packages (from multiqc) (0.18.2)
Requirement already satisfied: spectra>=0.0.10 in /home/linuxbrew/.linuxbrew/lib/python3.9/site-packages (from multiqc) (0.0.11)
Requirement already satisfied: markdown in /home/linuxbrew/.linuxbrew/lib/python3.9/site-packages (from multiqc) (3.4.1)
Requirement already satisfied: networkx>=2.5.1 in /home/linuxbrew/.linuxbrew/lib/python3.9/site-packages (from multiqc) (2.8.5)
Requirement already satisfied: numpy in /home/linuxbrew/.linuxbrew/lib/python3.9/site-packages (from multiqc) (1.22.4)
Requirement already satisfied: pyyaml>=4 in /home/linuxbrew/.linuxbrew/lib/python3.9/site-packages (from multiqc) (6.0)
Requirement already satisfied: rich>=10 in /home/linuxbrew/.linuxbrew/lib/python3.9/site-packages (from multiqc) (12.5.1)
Requirement already satisfied: matplotlib>=2.1.1 in /home/linuxbrew/.linuxbrew/lib/python3.9/site-packages (from multiqc) (3.5.2)
Requirement already satisfied: lzstring in /home/linuxbrew/.linuxbrew/lib/python3.9/site-packages (from multiqc) (1.0.4)
Requirement already satisfied: requests in /home/linuxbrew/.linuxbrew/lib/python3.9/site-packages (from multiqc) (2.28.1)
Requirement already satisfied: simplejson in /home/linuxbrew/.linuxbrew/lib/python3.9/site-packages (from multiqc) (3.17.6)
Requirement already satisfied: coloredlogs in /home/linuxbrew/.linuxbrew/lib/python3.9/site-packages (from multiqc) (15.0.1)
Requirement already satisfied: MarkupSafe>=2.0 in /home/linuxbrew/.linuxbrew/lib/python3.9/site-packages (from jinja2>=2.9->multiqc) (2.1.1)
Requirement already satisfied: cycler>=0.10 in /home/linuxbrew/.linuxbrew/lib/python3.9/site-packages (from matplotlib>=2.1.1->multiqc) (0.11.0)
Requirement already satisfied: kiwisolver>=1.0.1 in /home/linuxbrew/.linuxbrew/lib/python3.9/site-packages (from matplotlib>=2.1.1->multiqc) (1.4.4)
Requirement already satisfied: fonttools>=4.22.0 in /home/linuxbrew/.linuxbrew/lib/python3.9/site-packages (from matplotlib>=2.1.1->multiqc) (4.34.4)
Requirement already satisfied: packaging>=20.0 in /home/linuxbrew/.linuxbrew/lib/python3.9/site-packages (from matplotlib>=2.1.1->multiqc) (21.3)
Requirement already satisfied: pillow>=6.2.0 in /home/linuxbrew/.linuxbrew/lib/python3.9/site-packages (from matplotlib>=2.1.1->multiqc) (9.2.0)
Requirement already satisfied: pyparsing>=2.2.1 in /home/linuxbrew/.linuxbrew/lib/python3.9/site-packages (from matplotlib>=2.1.1->multiqc) (3.0.9)
Requirement already satisfied: python-dateutil>=2.7 in /home/linuxbrew/.linuxbrew/lib/python3.9/site-packages (from matplotlib>=2.1.1->multiqc) (2.8.2)
Requirement already satisfied: commonmark<0.10.0,>=0.9.0 in /home/linuxbrew/.linuxbrew/lib/python3.9/site-packages (from rich>=10->multiqc) (0.9.1)
Requirement already satisfied: pygments<3.0.0,>=2.6.0 in /home/linuxbrew/.linuxbrew/lib/python3.9/site-packages (from rich>=10->multiqc) (2.12.0)
Requirement already satisfied: colormath>=3.0.0 in /home/linuxbrew/.linuxbrew/lib/python3.9/site-packages (from spectra>=0.0.10->multiqc) (3.0.0)
Requirement already satisfied: humanfriendly>=9.1 in /home/linuxbrew/.linuxbrew/lib/python3.9/site-packages (from coloredlogs->multiqc) (10.0)
Requirement already satisfied: importlib-metadata>=4.4 in /home/linuxbrew/.linuxbrew/lib/python3.9/site-packages (from markdown->multiqc) (4.12.0)
Requirement already satisfied: idna<4,>=2.5 in /home/linuxbrew/.linuxbrew/lib/python3.9/site-packages (from requests->multiqc) (3.3)
Requirement already satisfied: charset-normalizer<3,>=2 in /home/linuxbrew/.linuxbrew/lib/python3.9/site-packages (from requests->multiqc) (2.1.0)
Requirement already satisfied: certifi>=2017.4.17 in /home/linuxbrew/.linuxbrew/lib/python3.9/site-packages (from requests->multiqc) (2022.6.15)
Requirement already satisfied: urllib3<1.27,>=1.21.1 in /home/linuxbrew/.linuxbrew/lib/python3.9/site-packages (from requests->multiqc) (1.26.10)
Requirement already satisfied: zipp>=0.5 in /home/linuxbrew/.linuxbrew/lib/python3.9/site-packages (from importlib-metadata>=4.4->markdown->multiqc) (3.8.1)
Requirement already satisfied: six>=1.5 in /home/linuxbrew/.linuxbrew/lib/python3.9/site-packages (from python-dateutil>=2.7->matplotlib>=2.1.1->multiqc) (1.16.0)
DEPRECATION: Configuring installation scheme with distutils config files is deprecated and will no longer work in the near future. If you are using a Homebrew or Linuxbrew Python, please see discussion at https://github.com/Homebrew/homebrew-core/issues/76621
--- Logging error ---
Traceback (most recent call last):
  File "/home/linuxbrew/.linuxbrew/opt/python@3.9/lib/python3.9/site-packages/pip/_internal/utils/logging.py", line 177, in emit
    self.console.print(renderable, overflow="ignore", crop=False, style=style)
  File "/home/linuxbrew/.linuxbrew/opt/python@3.9/lib/python3.9/site-packages/pip/_vendor/rich/console.py", line 1752, in print
    extend(render(renderable, render_options))
  File "/home/linuxbrew/.linuxbrew/opt/python@3.9/lib/python3.9/site-packages/pip/_vendor/rich/console.py", line 1390, in render
    for render_output in iter_render:
  File "/home/linuxbrew/.linuxbrew/opt/python@3.9/lib/python3.9/site-packages/pip/_internal/utils/logging.py", line 134, in __rich_console__
    for line in lines:
  File "/home/linuxbrew/.linuxbrew/opt/python@3.9/lib/python3.9/site-packages/pip/_vendor/rich/segment.py", line 245, in split_lines
    for segment in segments:
  File "/home/linuxbrew/.linuxbrew/opt/python@3.9/lib/python3.9/site-packages/pip/_vendor/rich/console.py", line 1368, in render
    renderable = rich_cast(renderable)
  File "/home/linuxbrew/.linuxbrew/opt/python@3.9/lib/python3.9/site-packages/pip/_vendor/rich/protocol.py", line 36, in rich_cast
    renderable = cast_method()
  File "/home/linuxbrew/.linuxbrew/opt/python@3.9/lib/python3.9/site-packages/pip/_internal/self_outdated_check.py", line 130, in __rich__
    pip_cmd = get_best_invocation_for_this_pip()
  File "/home/linuxbrew/.linuxbrew/opt/python@3.9/lib/python3.9/site-packages/pip/_internal/utils/entrypoints.py", line 58, in get_best_invocation_for_this_pip
    if found_executable and os.path.samefile(
  File "/home/linuxbrew/.linuxbrew/opt/python@3.9/lib/python3.9/genericpath.py", line 101, in samefile
    s2 = os.stat(f2)
FileNotFoundError: [Errno 2] No such file or directory: '/home/linuxbrew/.linuxbrew/opt/python@3.9/bin/pip'
Call stack:
  File "/home/linuxbrew/.linuxbrew/bin/pip", line 8, in <module>
    sys.exit(main())
  File "/home/linuxbrew/.linuxbrew/opt/python@3.9/lib/python3.9/site-packages/pip/_internal/cli/main.py", line 70, in main
    return command.main(cmd_args)
  File "/home/linuxbrew/.linuxbrew/opt/python@3.9/lib/python3.9/site-packages/pip/_internal/cli/base_command.py", line 101, in main
    return self._main(args)
  File "/home/linuxbrew/.linuxbrew/opt/python@3.9/lib/python3.9/site-packages/pip/_internal/cli/base_command.py", line 223, in _main
    self.handle_pip_version_check(options)
  File "/home/linuxbrew/.linuxbrew/opt/python@3.9/lib/python3.9/site-packages/pip/_internal/cli/req_command.py", line 148, in handle_pip_version_check
    pip_self_version_check(session, options)
  File "/home/linuxbrew/.linuxbrew/opt/python@3.9/lib/python3.9/site-packages/pip/_internal/self_outdated_check.py", line 237, in pip_self_version_check
    logger.info("[present-rich] %s", upgrade_prompt)
  File "/home/linuxbrew/.linuxbrew/opt/python@3.9/lib/python3.9/logging/__init__.py", line 1446, in info
    self._log(INFO, msg, args, **kwargs)
  File "/home/linuxbrew/.linuxbrew/opt/python@3.9/lib/python3.9/logging/__init__.py", line 1589, in _log
    self.handle(record)
  File "/home/linuxbrew/.linuxbrew/opt/python@3.9/lib/python3.9/logging/__init__.py", line 1599, in handle
    self.callHandlers(record)
  File "/home/linuxbrew/.linuxbrew/opt/python@3.9/lib/python3.9/logging/__init__.py", line 1661, in callHandlers
    hdlr.handle(record)
  File "/home/linuxbrew/.linuxbrew/opt/python@3.9/lib/python3.9/logging/__init__.py", line 952, in handle
    self.emit(record)
  File "/home/linuxbrew/.linuxbrew/opt/python@3.9/lib/python3.9/site-packages/pip/_internal/utils/logging.py", line 179, in emit
    self.handleError(record)
Message: '[present-rich] %s'
Arguments: (UpgradePrompt(old='22.1.2', new='22.2.2'),)
```
### 2.4 cutadapt  

pip install cutadapt 类似  
```
# Check that cutadapt is installed
xuruizhi@DESKTOP-HI65AUV:~$ cutadapt --version
4.1

# Check that FastQC is installed
xuruizhi@DESKTOP-HI65AUV:~$ fastqc -v
FastQC v0.11.9
```

### 2.5 质量修剪  
* Trim Galore
```
cd /mnt/d/biosoft
# 先挂载到d盘相应文件

wget https://github.com/FelixKrueger/TrimGalore/archive/0.6.3.tar.gz -O TrimGalore.gz

gzip -d TrimGalore.gz
```
! [作者GitHub](https://github.com/FelixKrueger/TrimGalore)已经更新至2021年7月的0.6.6版本
```
# Install Trim Galore
curl -fsSL https://github.com/FelixKrueger/TrimGalore/archive/0.6.6.tar.gz -o TrimGalore.tar.gz
tar xvzf TrimGalore.tar.gz

# Run Trim Galore
~/TrimGalore-0.6.6/trim_galore
```
结果：  
```
xuruizhi@DESKTOP-HI65AUV:/mnt/d/biosoft$ curl -fsSL https://github.com/FelixKrueger/TrimGalore/archive/0.6.6.tar.gz -o TrimGalore.tar.gz
xuruizhi@DESKTOP-HI65AUV:/mnt/d/biosoft$ ls
TrimGalore.tar.gz  Trimmomatic-0.38  Trimmomatic-0.38.zip  hisat2-2.2.1  sortmerna-2.1  sortmerna-2.1.tar.gz  wget-log
xuruizhi@DESKTOP-HI65AUV:/mnt/d/biosoft$ tar xvzf TrimGalore.tar.gz
TrimGalore-0.6.6/
TrimGalore-0.6.6/.travis.yml
TrimGalore-0.6.6/Changelog.md
TrimGalore-0.6.6/Docs/
TrimGalore-0.6.6/Docs/Images/
TrimGalore-0.6.6/Docs/Images/Thumbs.db
TrimGalore-0.6.6/Docs/Images/adapters.png
TrimGalore-0.6.6/Docs/Images/adapters_fixed.png
TrimGalore-0.6.6/Docs/Images/fixed_quals.png
TrimGalore-0.6.6/Docs/Images/fixed_quals_per_sequence.png
TrimGalore-0.6.6/Docs/Images/logo.png
TrimGalore-0.6.6/Docs/Images/pigz_bench.png
TrimGalore-0.6.6/Docs/Images/poor_q_per_sequence.png
TrimGalore-0.6.6/Docs/Images/poor_quals.png
TrimGalore-0.6.6/Docs/RRBS_Guide.pdf
TrimGalore-0.6.6/Docs/Trim_Galore_User_Guide.md
TrimGalore-0.6.6/LICENSE
TrimGalore-0.6.6/README.md
TrimGalore-0.6.6/test_files/
TrimGalore-0.6.6/test_files/4_seqs_with_Ns.fastq.gz
TrimGalore-0.6.6/test_files/clock_10K_R1.fastq.gz
TrimGalore-0.6.6/test_files/clock_10K_R2.fastq.gz
TrimGalore-0.6.6/test_files/colorspace_file.fastq
TrimGalore-0.6.6/test_files/empty_file.fastq
TrimGalore-0.6.6/test_files/illumina_100K.fastq.gz
TrimGalore-0.6.6/test_files/illumina_100K.fastq.gz_trimming_report.txt
TrimGalore-0.6.6/test_files/illumina_10K.fastq.gz
TrimGalore-0.6.6/test_files/illumina_10K.fastq.gz_trimming_report.txt
TrimGalore-0.6.6/test_files/illumina_10K_trimmed.fq.gz
TrimGalore-0.6.6/test_files/nextera_100K.fastq.gz
TrimGalore-0.6.6/test_files/nextera_100K.fastq.gz_trimming_report.txt
TrimGalore-0.6.6/test_files/smallRNA_100K.fastq.gz
TrimGalore-0.6.6/test_files/smallRNA_100K.fastq.gz_trimming_report.txt
TrimGalore-0.6.6/test_files/smallRNA_100K_R1.fastq.gz_trimming_report.txt
TrimGalore-0.6.6/test_files/smallRNA_100K_R2.fastq.gz
TrimGalore-0.6.6/test_files/smallRNA_100K_R2.fastq.gz_trimming_report.txt
TrimGalore-0.6.6/test_files/truncated.fq.gz
TrimGalore-0.6.6/trim_galore
```
`！师兄的办法会得到一个单独的TrimGalore文件；作者的办法会得到包括trim_galore及其license在内的一个文件夹`

* fastp  
* trimmomatic  
```
cd /mnt/d/biosoft
# 先挂载到d盘相应文件 

wget http://www.usadellab.org/cms/uploads/supplementary/Trimmomatic/Trimmomatic-0.38.zip
unzip Trimmomatic-0.38.zip

cd Trimmomatic-0.38

# 导入临时环境变量
export PATH="$(pwd):$PATH"
```
### 2.6 hisat2  

1. [hisat2官网更改](https://daehwankimlab.github.io/hisat2/)
2. 右侧download下载,直接点击下载即可，不需要回到终端再下载。下载完成后剪切到d/biosoft文件夹内解压
```
Version: HISAT2 2.2.1
Release Date: 7/24/2020

Linux_x86_64	https://cloud.biohpc.swmed.edu/index.php/s/oTtGWbWjaxsQ2Ho/download
```
3. 回到终端写入环境
```

# 导入临时环境变量
$ export PATH="~/biosoft/hisat2-2.1.0:$PATH"

# 测试是否可用
$ hisat2 -h

xuruizhi@DESKTOP-HI65AUV:/mnt/d/biosoft$  hisat2 -h
HISAT2 version 2.2.1 by Daehwan Kim (infphilo@gmail.com, www.ccb.jhu.edu/people/infphilo)
Usage:
```
### sortmerna [选做]
有时因为安全问题无法直接从网页下载，最好还是回到终端在本地下载  

```
cd /mnt/d/biosoft
# 先挂载到d盘相应文件 
# 下载软件,已经是最新版本
$ wget https://github.com/biocore/sortmerna/archive/2.1.tar.gz -O sortmerna-2.1.tar.gz

# 解压
$ tar -xzvf sortmerna-2.1.tar.gz
$ cd sortmerna-2.1

# 配置相关信息
$ ./configure --prefix=$PWD

# 编译
$ make -j 4

# 查看是否成功
$ ./sortmerna --help

# 导入到环境变量
$ export PATH="$(pwd):$PATH"
```
结果：  
```
解压后：
xuruizhi@DESKTOP-HI65AUV:/mnt/d/biosoft$  tar -xzvf sortmerna-2.1.tar.gz
sortmerna-2.1/
sortmerna-2.1/.gitignore
sortmerna-2.1/AUTHORS
sortmerna-2.1/COPYING
sortmerna-2.1/COPYING.LESSER.txt
sortmerna-2.1/ChangeLog
sortmerna-2.1/INSTALL
sortmerna-2.1/LICENSE.LESSER.txt
sortmerna-2.1/LICENSE.txt
sortmerna-2.1/Makefile.am
sortmerna-2.1/Makefile.in
sortmerna-2.1/NEWS
sortmerna-2.1/README
sortmerna-2.1/README.md
sortmerna-2.1/SortMeRNA-User-Manual-2.1.pdf
sortmerna-2.1/SortMeRNA-User-Manual.tex
sortmerna-2.1/_autotools_batch_.sh
sortmerna-2.1/aclocal.m4
sortmerna-2.1/alp/
sortmerna-2.1/alp/.deps/
sortmerna-2.1/alp/.deps/.dirstamp
sortmerna-2.1/alp/.dirstamp
sortmerna-2.1/alp/gumbel_params.cpp
sortmerna-2.1/alp/gumbel_params.hpp
sortmerna-2.1/alp/gumbelparams_app.t.cpp
sortmerna-2.1/alp/njn_approx.hpp
sortmerna-2.1/alp/njn_doubletype.hpp
sortmerna-2.1/alp/njn_dynprogprob.cpp
sortmerna-2.1/alp/njn_dynprogprob.hpp
sortmerna-2.1/alp/njn_dynprogproblim.cpp
sortmerna-2.1/alp/njn_dynprogproblim.hpp
sortmerna-2.1/alp/njn_dynprogprobproto.cpp
sortmerna-2.1/alp/njn_dynprogprobproto.hpp
sortmerna-2.1/alp/njn_function.hpp
sortmerna-2.1/alp/njn_integer.hpp
sortmerna-2.1/alp/njn_ioutil.cpp
sortmerna-2.1/alp/njn_ioutil.hpp
sortmerna-2.1/alp/njn_localmaxstat.cpp
sortmerna-2.1/alp/njn_localmaxstat.hpp
sortmerna-2.1/alp/njn_localmaxstatmatrix.cpp
sortmerna-2.1/alp/njn_localmaxstatmatrix.hpp
sortmerna-2.1/alp/njn_localmaxstatutil.cpp
sortmerna-2.1/alp/njn_localmaxstatutil.hpp
sortmerna-2.1/alp/njn_matrix.cpp
sortmerna-2.1/alp/njn_matrix.hpp
sortmerna-2.1/alp/njn_memutil.hpp
sortmerna-2.1/alp/njn_random.cpp
sortmerna-2.1/alp/njn_random.hpp
sortmerna-2.1/alp/njn_root.hpp
sortmerna-2.1/alp/njn_stringutil.cpp
sortmerna-2.1/alp/njn_stringutil.hpp
sortmerna-2.1/alp/njn_uniform.hpp
sortmerna-2.1/alp/njn_vector.hpp
sortmerna-2.1/alp/random_param.txt
sortmerna-2.1/alp/sls_alp.cpp
sortmerna-2.1/alp/sls_alp.hpp
sortmerna-2.1/alp/sls_alp_data.cpp
sortmerna-2.1/alp/sls_alp_data.hpp
sortmerna-2.1/alp/sls_alp_regression.cpp
sortmerna-2.1/alp/sls_alp_regression.hpp
sortmerna-2.1/alp/sls_alp_sim.cpp
sortmerna-2.1/alp/sls_alp_sim.hpp
sortmerna-2.1/alp/sls_normal_distr_array.hpp
sortmerna-2.1/alp/sls_pvalues.cpp
sortmerna-2.1/alp/sls_pvalues.hpp
sortmerna-2.1/build.sh
sortmerna-2.1/cmph/
sortmerna-2.1/cmph/.deps/
sortmerna-2.1/cmph/.deps/.dirstamp
sortmerna-2.1/cmph/.dirstamp
sortmerna-2.1/cmph/bdz.c
sortmerna-2.1/cmph/bdz.h
sortmerna-2.1/cmph/bdz_ph.c
sortmerna-2.1/cmph/bdz_ph.h
sortmerna-2.1/cmph/bdz_structs.h
sortmerna-2.1/cmph/bdz_structs_ph.h
sortmerna-2.1/cmph/bitbool.h
sortmerna-2.1/cmph/bmz.c
sortmerna-2.1/cmph/bmz.h
sortmerna-2.1/cmph/bmz8.c
sortmerna-2.1/cmph/bmz8.h
sortmerna-2.1/cmph/bmz8_structs.h
sortmerna-2.1/cmph/bmz_structs.h
sortmerna-2.1/cmph/brz.c
sortmerna-2.1/cmph/brz.h
sortmerna-2.1/cmph/brz_structs.h
sortmerna-2.1/cmph/buffer_entry.c
sortmerna-2.1/cmph/buffer_entry.h
sortmerna-2.1/cmph/buffer_manager.c
sortmerna-2.1/cmph/buffer_manager.h
sortmerna-2.1/cmph/chd.c
sortmerna-2.1/cmph/chd.h
sortmerna-2.1/cmph/chd_ph.c
sortmerna-2.1/cmph/chd_ph.h
sortmerna-2.1/cmph/chd_structs.h
sortmerna-2.1/cmph/chd_structs_ph.h
sortmerna-2.1/cmph/chm.c
sortmerna-2.1/cmph/chm.h
sortmerna-2.1/cmph/chm_structs.h
sortmerna-2.1/cmph/cmph.c
sortmerna-2.1/cmph/cmph.h
sortmerna-2.1/cmph/cmph_structs.c
sortmerna-2.1/cmph/cmph_structs.h
sortmerna-2.1/cmph/cmph_time.h
sortmerna-2.1/cmph/cmph_types.h
sortmerna-2.1/cmph/compressed_rank.c
sortmerna-2.1/cmph/compressed_rank.h
sortmerna-2.1/cmph/compressed_seq.c
sortmerna-2.1/cmph/compressed_seq.h
sortmerna-2.1/cmph/debug.h
sortmerna-2.1/cmph/fch.c
sortmerna-2.1/cmph/fch.h
sortmerna-2.1/cmph/fch_buckets.c
sortmerna-2.1/cmph/fch_buckets.h
sortmerna-2.1/cmph/fch_structs.h
sortmerna-2.1/cmph/graph.c
sortmerna-2.1/cmph/graph.h
sortmerna-2.1/cmph/hash.c
sortmerna-2.1/cmph/hash.h
sortmerna-2.1/cmph/hash_state.h
sortmerna-2.1/cmph/jenkins_hash.c
sortmerna-2.1/cmph/jenkins_hash.h
sortmerna-2.1/cmph/miller_rabin.c
sortmerna-2.1/cmph/miller_rabin.h
sortmerna-2.1/cmph/select.c
sortmerna-2.1/cmph/select.h
sortmerna-2.1/cmph/select_lookup_tables.h
sortmerna-2.1/cmph/vqueue.c
sortmerna-2.1/cmph/vqueue.h
sortmerna-2.1/cmph/vstack.c
sortmerna-2.1/cmph/vstack.h
sortmerna-2.1/compile
sortmerna-2.1/configure
sortmerna-2.1/configure.ac
sortmerna-2.1/depcomp
sortmerna-2.1/include/
sortmerna-2.1/include/bitvector.hpp
sortmerna-2.1/include/common.hpp
sortmerna-2.1/include/config.h
sortmerna-2.1/include/config.h.in
sortmerna-2.1/include/indexdb.hpp
sortmerna-2.1/include/loadgenome.hpp
sortmerna-2.1/include/memcheck.h
sortmerna-2.1/include/minoccur.hpp
sortmerna-2.1/include/outputformats.hpp
sortmerna-2.1/include/paralleltraversal.hpp
sortmerna-2.1/include/ssw.h
sortmerna-2.1/include/stamp-h1
sortmerna-2.1/include/triestats.hpp
sortmerna-2.1/index/
sortmerna-2.1/index/.tobuild
sortmerna-2.1/indexdb_rna.1
sortmerna-2.1/install-sh
sortmerna-2.1/missing
sortmerna-2.1/rRNA_databases/
sortmerna-2.1/rRNA_databases/README.txt
sortmerna-2.1/rRNA_databases/rfam-5.8s-database-id98.fasta
sortmerna-2.1/rRNA_databases/rfam-5s-database-id98.fasta
sortmerna-2.1/rRNA_databases/scripts/
sortmerna-2.1/rRNA_databases/scripts/edit_U_to_T_rna.py
sortmerna-2.1/rRNA_databases/scripts/generate_databases.sh
sortmerna-2.1/rRNA_databases/scripts/hmmer_clean.sh
sortmerna-2.1/rRNA_databases/silva-arc-16s-id95.fasta
sortmerna-2.1/rRNA_databases/silva-arc-23s-id98.fasta
sortmerna-2.1/rRNA_databases/silva-bac-16s-id90.fasta
sortmerna-2.1/rRNA_databases/silva-bac-23s-id98.fasta
sortmerna-2.1/rRNA_databases/silva-euk-18s-id95.fasta
sortmerna-2.1/rRNA_databases/silva-euk-28s-id98.fasta
sortmerna-2.1/rRNA_databases/silva_ids_acc_tax.tar.gz
sortmerna-2.1/scripts/
sortmerna-2.1/scripts/merge-paired-reads.sh
sortmerna-2.1/scripts/unmerge-paired-reads.sh
sortmerna-2.1/sortmerna.1
sortmerna-2.1/src/
sortmerna-2.1/src/.deps/
sortmerna-2.1/src/.deps/.dirstamp
sortmerna-2.1/src/.dirstamp
sortmerna-2.1/src/bitvector.cpp
sortmerna-2.1/src/indexdb.cpp
sortmerna-2.1/src/main.cpp
sortmerna-2.1/src/minoccur.cpp
sortmerna-2.1/src/outputformats.cpp
sortmerna-2.1/src/paralleltraversal.cpp
sortmerna-2.1/src/ssw.c
sortmerna-2.1/src/ssw_example.c
sortmerna-2.1/tests/
sortmerna-2.1/tests/data/
sortmerna-2.1/tests/data/empty_file.fasta
sortmerna-2.1/tests/data/gg_13_8_ref_set.fasta
sortmerna-2.1/tests/data/illumina_GQ099317.fasta
sortmerna-2.1/tests/data/ref_GQ099317_forward_and_rc.fasta
sortmerna-2.1/tests/data/ref_short_seqs.fasta
sortmerna-2.1/tests/data/set2_environmental_study_550_amplicon.fasta
sortmerna-2.1/tests/data/set4_mate_pairs_metatranscriptomics.fastq
sortmerna-2.1/tests/data/set5_simulated_amplicon_silva_bac_16s.fasta
sortmerna-2.1/tests/data/set7_arc_bac_16S_database_match.fasta
sortmerna-2.1/tests/data/silva-arc-16s-database-id95.fasta
sortmerna-2.1/tests/data/silva-bac-16s-database-id85.fasta
sortmerna-2.1/tests/test_sortmerna.py
```
原本数据存放在`d/biosoft/sortmerna-2.1`文件夹内，移动数据库位置之前先在d盘新建一个database/sortmerna_db/rRNA_databases存放数据

```
xuruizhi@DESKTOP-HI65AUV:/mnt/d$ mkdir database
xuruizhi@DESKTOP-HI65AUV:/mnt/d$ cd database
。。。
xuruizhi@DESKTOP-HI65AUV:/mnt/d/database/sortmerna_db$ ls
rRNA_databases
```
```
# 把数据库文件移动到能找到的地方,d盘的database文件夹内
#没搞明白
$ mv ./rRNA_databases/ /mnt/d/database/sortmerna_db/rRNA_databases

# 相关库文件
$ cd ~/database/rRNA_databases
$ sortmerna_ref_data=$(pwd)/rRNA_databases/silva-bac-16s-id90.fasta,$(pwd)/index/silva-bac-16s-db:\
$(pwd)/rRNA_databases/silva-bac-23s-id98.fasta,$(pwd)/index/silva-bac-23s-db:\
$(pwd)/rRNA_databases/silva-arc-16s-id95.fasta,$(pwd)/index/silva-arc-16s-db:\
$(pwd)/rRNA_databases/silva-arc-23s-id98.fasta,$(pwd)/index/silva-arc-23s-db:\
$(pwd)/rRNA_databases/silva-euk-18s-id95.fasta,$(pwd)/index/silva-euk-18s-db:\
$(pwd)/rRNA_databases/silva-euk-28s-id98.fasta,$(pwd)/index/silva-euk-28s-db:\
$(pwd)/rRNA_databases/rfam-5s-database-id98.fasta,$(pwd)/index/rfam-5s-db:\
$(pwd)/rRNA_databases/rfam-5.8s-database-id98.fasta,$(pwd)/index/rfam-5.8s-db

# 真核生物的rRNA不需要那么多(5s, 5.8s, 18s, 28s)
$ euk_rNRA_ref_data=$(pwd)/rRNA_databases/silva-euk-18s-id95.fasta,$(pwd)/index/silva-euk-18s-db:\
$(pwd)/rRNA_databases/silva-euk-28s-id98.fasta,$(pwd)/index/silva-euk-28s-db:\
$(pwd)/rRNA_databases/rfam-5s-database-id98.fasta,$(pwd)/index/rfam-5s-db:\
$(pwd)/rRNA_databases/rfam-5.8s-database-id98.fasta,$(pwd)/index/rfam-5.8s-db

# 建立数据库索引
$ indexdb_rna --ref $data
```

### 2.7 samtools
最新版本为1.16  
本地下载时，在配制这步出错，使用`brew install samtools`安装

### 2.8 HTseq
```
pip install -i https://pypi.tuna.tsinghua.edu.cn/simple HTseq
```

### 2.9 R
最新版本4.2.1_2  
先进入官网，用清华镜像源下载合适版本的R，再`brew install r`

```
xuruizhi@DESKTOP-HI65AUV:~$ brew install r
HOMEBREW_BREW_GIT_REMOTE set: using https://mirrors.tuna.tsinghua.edu.cn/git/homebrew/brew.git for Homebrew/brew Git remote.
HOMEBREW_CORE_GIT_REMOTE set: using https://mirrors.tuna.tsinghua.edu.cn/git/homebrew/homebrew-core.git for Homebrew/core Git remote.
Running `brew update --auto-update`...
Warning: r 4.2.1_2 is already installed and up-to-date.
To reinstall 4.2.1_2, run:
  brew reinstall r
  ```

  ### 2.10 Rstudio
进入网站：`https://www.rstudio.com/products/rstudio/download/`  
R studio 可以在 Windows 下安装;
选择版本下载,下载完成之后双击安装。  
`
Ubuntu 18+/Debian 10+	 rstudio-2022.07.1-554-amd64.deb
`